### PR TITLE
kernel: gpio-nct5104d: use devm

### PR DESCRIPTION
**Pull request by neheb**
Allows removal of the remove function by letting devm handle everything.

Avoids gotos as well.

### CODE REVIEW

**Comment by neheb** [comment]
Interestingly enough, a driver was attempted to be upstreamed: https://lore.kernel.org/lkml/20220712143237.13992-1-henning.schild@siemens.com/T/

Wonder if it makes sense to use that instead of this. I don't have an APU.

**Comment by neheb** [comment]
Looking closer at this, I wonder if this is enough to replace this driver...

```
--- a/drivers/gpio/gpio-f7188x.c
+++ b/drivers/gpio/gpio-f7188x.c
@@ -48,7 +48,11 @@
 /*
  * Nuvoton devices.
  */
+#define SIO_NCT5104D_ID                0x1061  /* NCT5104D chipset ID */
+#define SIO_NCT6106D_ID                0xC452  /* NCT6106D chipset ID */
+#define SIO_NCT6116D_ID                0xD282  /* NCT6116D chipset ID */
 #define SIO_NCT6126D_ID                0xD283  /* NCT6126D chipset ID */
+#define SIO_NCT6122D_ID                0xD2A3  /* NCT6122D chipset ID */
 
 #define SIO_LD_GPIO_NUVOTON    0x07    /* GPIO logical device */
 
@@ -62,7 +66,11 @@ enum chips {
        f81866,
        f81804,
        f81865,
+       nct5104d,
+       nct6106d,
+       nct6116d,
        nct6126d,
+       nct6122d,
 };
 
 static const char * const f7188x_names[] = {
@@ -74,7 +82,11 @@ static const char * const f7188x_names[] = {
        "f81866",
        "f81804",
        "f81865",
+       "nct5104d",
+       "nct6106d",
+       "nct6116d",
        "nct6126d",
+       "nct6122d",
 };
 
 struct f7188x_sio {
@@ -559,10 +571,26 @@ static int __init f7188x_find(int addr, struct f7188x_sio *sio)
        case SIO_F81865_ID:
                sio->type = f81865;
                break;
+       case SIO_NCT5104D_ID:
+               sio->device = SIO_LD_GPIO_NUVOTON;
+               sio->type = nct5104d;
+               break;
+       case SIO_NCT6106D_ID:
+               sio->device = SIO_LD_GPIO_NUVOTON;
+               sio->type = nct6106d;
+               break;
+       case SIO_NCT6116D_ID:
+               sio->device = SIO_LD_GPIO_NUVOTON;
+               sio->type = nct6116d;
+               break;
        case SIO_NCT6126D_ID:
                sio->device = SIO_LD_GPIO_NUVOTON;
                sio->type = nct6126d;
                break;
+       case SIO_NCT6122D_ID:
+               sio->device = SIO_LD_GPIO_NUVOTON;
+               sio->type = nct6122d;
+               break;
        default:
                pr_info("Unsupported Fintek device 0x%04x\n", devid);
                goto err;

```

**Comment by neheb** [comment]
ping @TimelessNL @riptidewave93

**Review by riptidewave93 — APPROVED** [review]
LGTM as long as you tested to confirm it works as expected.

File: package/kernel/gpio-nct5104d/src/gpio-nct5104d.c [hunk 1]
```diff
@@ -270,45 +270,17 @@ static int nct5104d_gpio_probe(struct platform_device *pdev)
...
-static int nct5104d_gpio_remove(struct platform_device *pdev)
-{
-	int i;
-	struct nct5104d_gpio_data *data = platform_get_drvdata(pdev);
-
-	for (i = 0; i < data->nr_bank; i++) {
-		struct nct5104d_gpio_bank *bank = &data->bank[i];
-
-		gpiochip_remove (&bank->chip);
+		err = devm_gpiochip_add_data(&pdev->dev, &bank->chip, bank);
```
Comment (hauke) [inline package/kernel/gpio-nct5104d/src/gpio-nct5104d.c:280]: Why do you set data?
I think it should look like this:
```suggestion
		err = devm_gpiochip_add_data(&pdev->dev, &bank->chip, NULL);
```

Comment (neheb) [reply package/kernel/gpio-nct5104d/src/gpio-nct5104d.c:280]: what's wrong with setting data?

Comment (neheb) [reply package/kernel/gpio-nct5104d/src/gpio-nct5104d.c]: Changed to NULL.

Comment (neheb) [reply package/kernel/gpio-nct5104d/src/gpio-nct5104d.c:276]: Ah nvm, I see the issue. gpiochip_get_data was missing.

Comment (hauke) [reply package/kernel/gpio-nct5104d/src/gpio-nct5104d.c:276]: This was not wrong, but the old code did not set the data attribute and it is also not used.

**Pull request merged**